dsp/modules/tensorrt_llm.py

```diff
@@ -0,0 +1,229 @@
+            self.model_name_or_path,
+            legacy=False,
+            padding_side="left",
+            truncation_side="left",
+            trust_remote_code=True,
+            use_fast=True,
+        )
+
+        self.pad_id = (
+            self.tokenizer.eos_token_id if self.tokenizer.pad_token_id is None else self.tokenizer.pad_token_id
+        )
+        self.end_id = self.tokenizer.eos_token_id
+
+        # Configure TensorRT
+        self.runtime_rank = tensorrt_llm.mpi_rank()
+        self.runner, self._runner_kwargs = load_tensorrt_model(engine_dir=self.engine_dir, **engine_kwargs)
+        self.history = []
+
+    def _generate(self, prompt: Union[list[dict[str, str]], str], **kwargs) -> list[str]:
+        import torch
+
+        input_ids = tokenize(prompt=prompt, tokenizer=self.tokenizer, **kwargs)
+        input_ids = torch.tensor(input_ids, dtype=torch.int32)
+
+        run_kwargs = {}
+        defaults = {
+            "max_new_tokens": 1024,
+            "max_attention_window_size": None,
+            "sink_token_length": None,
+            "end_id": self.end_id,
+            "pad_id": self.pad_id,
+            "temperature": 1.0,
+            "top_k": 1,
+            "top_p": 0.0,
+            "num_beams": 1,
+            "length_penalty": 1.0,
+            "early_stopping": 1,
+            "repetition_penalty": 1.0,
+            "presence_penalty": 0.0,
+            "frequency_penalty": 0.0,
+            "stop_words_list": None,
+            "bad_words_list": None,
+            "streaming": False,
+            "return_dict": True,
+            "output_log_probs": False,
+            "output_cum_log_probs": False,
+            "output_sequence_lengths": True,
+        }
+
+        for k, v in defaults.items():
+            run_kwargs[k] = kwargs.get(k, v)
+
+        with torch.no_grad():
+            outputs = self.runner.generate(input_ids, **run_kwargs)
+        input_lengths = [x.size(0) for x in input_ids]
+
+        output_ids, sequence_lengths = outputs["output_ids"], outputs["sequence_lengths"]
+
+        # In case of current version of dspy it will always stay as 1
+        batch_size, num_beams, _ = output_ids.size()
+        batch_idx, beams = 0, []
+
+        for beam in range(num_beams):
+            output_begin = input_lengths[batch_idx]
+            output_end = sequence_lengths[batch_idx][beam]
+            outputs = output_ids[batch_idx][beam][output_begin:output_end].tolist()
+            output_text = self.tokenizer.decode(outputs)
+            beams.append(output_text)
+
+        return beams, run_kwargs
+
+    def basic_request(self, prompt, **kwargs) -> list[str]:
+        raw_kwargs = kwargs
+        response, all_kwargs = self._generate(prompt, **kwargs)
+        history = {
+            "prompt": prompt,
+            "response": response,
+            "raw_kwargs": raw_kwargs,
+            "kwargs": all_kwargs,
+        }
+        self.history.append(history)
```

Comment:
Just checking if this addition needs any updates to the [history parsing](https://github.com/stanfordnlp/dspy/blob/2a008d65b155d36d6e587aa9d4a10948cd031039/dsp/modules/lm.py#L34) as well - some of the other providers process the response differently

Reply:
Ohh yeah, I guess have't update that, let me add that too

Reply:
I updated it, let me know if this make more sense now. Thanks